Replace Action monad with Cli monad

mitchellwrosen opened the pull request:
## Overview

Continuation of work on #3253

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/unisonweb/unison/3281)
<!-- Reviewable:end -->

mitchellwrosen committed:
move some code around

mitchellwrosen committed:
add stub RunCli command

mitchellwrosen committed:
start moving LoopState into an IORef in the env, but now it's time to eat a sausage

mitchellwrosen committed:
continue work on converting Action to Cli

mitchellwrosen committed:
finish refactoring handleCreatePullRequest

mitchellwrosen committed:
convert auth.login to Cli monad

mitchellwrosen committed:
move handleDependents to Cli monad

mitchellwrosen committed:
begin porting pr.load to Cli monad

tstat committed:
add get/put loop state functions for cli

tstat committed:
handlePushToUnisonShare to Cli

tstat committed:
handleGist to Cli

tstat committed:
DebugDoctor to Cli

tstat committed:
DebugClearWatchI to Cli

tstat committed:
DebugNumberedArgsI to Cli

tstat committed:
namespaceDependencies -> Cli

mitchellwrosen committed:
move ResetRootI to the Cli monad

tstat committed:
currentPrettyPrintEnvDecl -> Cli

tstat committed:
namespaceDependenciesI -> Cli

mitchellwrosen committed:
move MoveBranchI to Cli monad

mitchellwrosen committed:
MovePatchI to Cli

mitchellwrosen committed:
CopyPatchI to Cli

mitchellwrosen committed:
DeletePatchI to Cli

mitchellwrosen committed:
DeleteBranchI to Cli

mitchellwrosen committed:
ApiI to Cli

mitchellwrosen committed:
FindI to Cli

mitchellwrosen committed:
ShowReflogI to Cli

mitchellwrosen committed:
ForkLocalBranchI to Cli

mitchellwrosen committed:
MergeLocalBranchI to Cli

mitchellwrosen committed:
fill in mergeBranchAndPropagateDefaultPatchCli

mitchellwrosen committed:
PreviewMergeLocalBranchI to Cli

mitchellwrosen committed:
DiffNamespaceI to Cli

mitchellwrosen committed:
SwitchBranchI to Cli

mitchellwrosen committed:
UpI to Cli

mitchellwrosen committed:
PopBranchI to Cli

tstat committed:
ListDependenciesI -> Cli

mitchellwrosen committed:
HistoryI to Cli

mitchellwrosen committed:
UndoI to Cli

mitchellwrosen committed:
UII, DocsToHtmlI to Cli

mitchellwrosen committed:
AliasTermI to Cli

mitchellwrosen committed:
AliasTypeI to Cli

tstat committed:
handlePushRemoteBranch -> Cli

mitchellwrosen committed:
AliasManyI to Cli

mitchellwrosen committed:
NamesI to Cli

tstat committed:
unhashTypeComponent -> Cli

tstat committed:
typecheckFile -> Cli

tstat committed:
applyPropagate doesn't need to be in Action

tstat committed:
genInitialCtorMapping -> Cli

tstat committed:
generalize unhashTermComponent

tstat committed:
verifyTermComponent -> Cli

tstat committed:
doTerm -> Cli

tstat committed:
doType -> Cli

tstat committed:
collectEdits -> Cli

tstat committed:
propagate -> Cli

tstat committed:
propagateAndApply -> Cli

mitchellwrosen committed:
LinkI/UnlinkI to Cli

mitchellwrosen committed:
getLinks' to Cli

mitchellwrosen committed:
getLinks to Cli

mitchellwrosen committed:
LinksI to Cli

tstat committed:
add stepManyAtCli' and stepAtCli'

tstat committed:
propagatePatch -> Cli

tstat committed:
add stepAtCliNoSync

tstat committed:
propagatePatchNoSync -> Cli

mitchellwrosen committed:
DocsI/DisplayI to Cli

mitchellwrosen committed:
cleanup using Cli.{get,put,modify}LoopState

mitchellwrosen committed:
CreateAuthorI to Cli

mitchellwrosen committed:
MoveTermI to Cli

mitchellwrosen committed:
MoveTypeI to Cli

mitchellwrosen committed:
delete to Cli

mitchellwrosen committed:
ShowDefinitionI to Cli

mitchellwrosen committed:
FindPatchI to Cli

mitchellwrosen committed:
FindShallowI to Cli

mitchellwrosen committed:
ResolveTypeNameI to Cli

mitchellwrosen committed:
fill in body of loadPropagateDiffDefaultPatchCli

mitchellwrosen committed:
fill in body of syncRootCli

tstat committed:
PullRemoteBranchI -> Cli

mitchellwrosen committed:
ResolveTermNameI to Cli

mitchellwrosen committed:
delete old resolveConfiguredUrl

mitchellwrosen committed:
ReplaceI to Cli

mitchellwrosen committed:
LoadI to Cli

mitchellwrosen committed:
AddI to Cli

mitchellwrosen committed:
UpdateI to Cli

mitchellwrosen committed:
PreviewAddI to Cli

mitchellwrosen committed:
PreviewUpdateI to Cli

mitchellwrosen committed:
TodoI to Cli

mitchellwrosen committed:
TestI to Cli

mitchellwrosen committed:
PropagatePatchI to Cli

mitchellwrosen committed:
ExecuteI to Cli

mitchellwrosen committed:
loadUnisonFile to Cli

mitchellwrosen committed:
MakeStandaloneI to Cli

mitchellwrosen committed:
IOTestI to Cli

mitchellwrosen committed:
MergeBuiltinsI to Cli

mitchellwrosen committed:
MergeIOBuiltinsI to Cli

mitchellwrosen committed:
ListEditsI to Cli

mitchellwrosen committed:
refactoring

mitchellwrosen committed:
DebugTypecheckedUnisonFileI to Cli

mitchellwrosen committed:
DebugDumpNamespacesI to Cli

mitchellwrosen committed:
DebugDumpNamespaceSimpleI to Cli

mitchellwrosen committed:
doRemoveReplacement to Cli

mitchellwrosen committed:
delete Action monad

mitchellwrosen committed:
delete Unison.Util.Free

mitchellwrosen committed:
make returnEarly take an Output

mitchellwrosen committed:
getMaybePatchAt

mitchellwrosen committed:
cleanup

mitchellwrosen committed:
more cleanup

mitchellwrosen committed:
getRootBranch / getRootBranch0

mitchellwrosen committed:
refactoring/cleanup in HandleInput

mitchellwrosen committed:
expectLatestTypecheckedFile

mitchellwrosen committed:
more refactoring/cleanup in HandleInput

mitchellwrosen committed:
refactor: reduce scope of root0 variable

mitchellwrosen committed:
refactor: bring defaultPatchPath to top-level

mitchellwrosen committed:
refactor: reduce scope of currentBranch0

mitchellwrosen committed:
refactor: reduce scope of latestFile' variable

mitchellwrosen committed:
refactor: reduce scope of hqLength variable

mitchellwrosen committed:
refactor: reduce scope of sbhLength variable

mitchellwrosen committed:
refactor: reduce scope of root' variable

mitchellwrosen committed:
whittle away the use of a couple more variables with huge scope

mitchellwrosen committed:
refactor: make manageLinks do less duplicate work

mitchellwrosen committed:
eliminate a use of resolveToAbsolute

mitchellwrosen committed:
work on Unison.Monad.Cli docs a bit

tstat committed:
cli docs

mitchellwrosen committed:
fix export

mitchellwrosen committed:
refactor: reduce scope of resolveSplit'

mitchellwrosen committed:
refactor: get codebase from environment in propagateAndApply

mitchellwrosen committed:
move some code around

mitchellwrosen committed:
refactor: remove a couple args from propagate, get them from env instead

mitchellwrosen committed:
refactor: fill in a few partial type signatures

mitchellwrosen committed:
document Unison.Cli.MonadUtils

mitchellwrosen committed:
finish clearing out Unison.Codebase.Editor.Command

mitchellwrosen committed:
permute ContT/ReaderT in Cli definition

mitchellwrosen committed:
rename HaltStep to Continue

mitchellwrosen committed:
refactor: delete uses of currentPath' and resolveToAbsolute in input
loop, define inputDescription at top level

mitchellwrosen committed:
Fix bug in move.term, move.type

tstat committed:
Fix bug in DeleteBranchI

tstat committed:
Merge remote-tracking branch 'origin/trunk' into unison-cli-tech-debt-2

pchiusano commented:
wow, an epic refactoring. looks a lot cleaner!

ChrisPenner commented on unison-cli/src/Unison/Codebase/Editor/HandleInput/AuthLogin.hs:
```diff
@@ -1,28 +1,218 @@
+          Right a -> Right a
+      | otherwise -> pure $ Left (InvalidTokenResponse tokenEndpoint $ "Received " <> tShow status <> " response from token endpoint")
+
+-- authenticateCodeserverCli :: CodeserverURI -> Cli r ()
```
What's going on here? I assumed this was just a transplanted comment, but this string doesn't show up on trunk at all 🤔

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadIO,
```
Is the lack of MonadUnliftIO going to be a problem?

It'd be great to have, but I assume it's somehow incompatible with the short-circuiting? Or is there some other reason?

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+-- * It is a reader monad: get the 'Env' environment with 'ask'.
+--
+-- * It is a state monad (kind of): get the 'LoopState' with 'getLoopState', and put it with 'putLoopState'. The state
+-- is stored in an @IORef@, so state changes are visible across threads, and persist even when an exception is raised.
```
This makes me a little nervous, I think it just means we'll have to be very careful about getting the root codebase in threads and such since it could change at any point without notice.

If an IORef allows us to use MonadUnliftIO though that might be nice?

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+      MonadIO,
+      MonadReader Env
+    )
+    via ReaderT Env (ContT (ReturnType r) IO)
```
I'm curious why `deriving via` rather than using  `ReaderT Env (ContT (ReturnType r) IO)` in the newtype directly?

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+    )
+    via ReaderT Env (ContT (ReturnType r) IO)
+
+-- | What a Cli action returns: a value, an instruction to continue processing input, or an instruction to stop
```
Should probably clarify that `continue processing input` means short-circuiting the current user command and jumping straight to the next input, as it's written it's a bit fuzzy about whether it continues execution or skips things.

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+instance
+  {-# OVERLAPS #-}
+  Functor f =>
+  IsLabel "currentPath" ((Path.Absolute -> f Path.Absolute) -> (LoopState -> f LoopState))
```
What was the rationale for using an overloaded label here as opposed to a regular ol' lens-as-a-value?

Is there some additional benefit? Otherwise I'd personally prefer to keep things simple and just use a regular lens 😄 , especially since we've already got a sprinkling of _both_ template haskell lenses and generic-lenses; adding another type might be confusing.

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+  Env {loopStateRef} <- ask
+  liftIO (writeIORef loopStateRef newSt)
+
+-- | Modify the loop state (not thread-safe).
```
I'm curious, what's the benefit of using an IORef instead of State if we're not aiming to use multiple threads or have MonadUnliftIO? 

Seems like it could be a foot-gun in the future.

If it's due to weird interactions between ContT and State I agree those are a pain, but I think we could work through them 🤔

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+  | LoadError
+  | LoadSuccess Text
+
+-- | Lift an action of type @IO (Either e a)@, given a continuation for @e@.
```
Does this combinator provide anything special over simply running the action and pattern-matching on the result?

I.e. are these two blocks roughly equivalent?

```
r <- liftIO action >>= \case
    Left e -> recover e
    Right a -> pure a
doSomething r

vs

ioE action recover >>= doSomething
```

?

I ask because I wouldn't expect to need to `unCli` and such to do this sort of thing; seems more complex that necessary so I'm guessing I'm missing something 🤔

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+short :: ReturnType r -> Cli r a
+short r = Cli \_k _env -> pure r
+
+-- | Short-circuit with a value. Returns @r@ to the nearest enclosing
```
Is this ever intended to be used outside of a `newBlock`?
If so, what's the `r` value used for and when should I use this instead of `returnEarlyWithoutOutput`?

ChrisPenner commented on unison-cli/src/Unison/Cli/Monad.hs:
```diff
@@ -0,0 +1,349 @@
+      Left err -> unCli (errK err) env k
+      Right val -> pure val
+
+-- | Run the given action in a new block, which delimits the scope of any 'succeedWith', 'with'/'withE', and 'time'
```
I think the fact that `newBlock` is the delimiter for all of `with`, `succeedWith` AND `time` might be a bit of a foot-gun;

E.g. what if I have a block like this:

```
do
  scratchFile <- with (withFile "scratch.u")
  doSomethingTimeIntensive
  useFile scratchFile
```

Then I want to time a block, I could see myself making this mistake (probably not in such a simple case, but in a real-world code situation it'd be easier to miss):

```
do
  newBlock do
    scratchFile <- with (withFile "scratch.u")
    doSomethingTimeIntensive
  useFile scratchFile
```

Now we free the file before using it :'( ;

You could make the same mistake using nested do-blocks too by returning the file outside of its scope, but it's considerably less likely IMO.

ChrisPenner reviewed (commented):
I love the ability to return-early and `continue`, but I have some concerns on the straight-line syntax for timing and resource acquisition.

I think this style separates the actual scoping of resource acquisition and timing from the place where they're acquired/started, which means it's really tough to tell what a given `newBlock`'s purpose is, but also it's easy to accidentally affect the scope of a resource or timing by adding a new `newBlock` somewhere for an entirely different purpose. 

Admittedly it's maybe more nesting to use do-blocks with `with*` combinators that take lambdas, but I that nesting clearly denotes control-flow and that is, in my opinion, definitely worth the syntactic overhead.

I personally noticed when reviewing some of the rewrites that I'd see a `newBlock` and not know why it's there; is it delimiting resource usage for some resource deep within one of the actions it's calling? Is it just there for timing? Can it be removed or will that break something? I don't ever have those questions when using explicit timing or resource acquisition blocks.

If one has the assumption that the next highest block is the cli-loop one might allocate a resource with the expectation it will live until the end of the command, but if anyone later wraps that allocation with a `newBlock` for some other allocation or even just for debugging something, it might break everything unexpectedly 😬 


If we still want block-scoped short-circuit mechanism I'd be much more comfortable with something like this:

```
do
  ...
  newBlock \returnEarly -> do
    case blah of
      1 -> returnEarly "something"
      2 -> returnEarly "something else"
      _ -> pure ()
    doSomethingElse
    pure "this is returned if you haven't returned early yet"
```

That way you can just jump-to-definition on `returnEarly` to see where you're actually returning to;
CPS-based early returns like this is really similar to using go-tos, so it's really nice to have the "returnEarly" as a label of sorts.

For resource acquisition I'd prefer to stick with nesting rather than straight-line syntax, i.e. use `withThing \thing -> do ...`
instead of `thing <- withThing`

ChrisPenner commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -582,397 +383,389 @@ loop e = do
+                      Cli.ioE (Codebase.importRemoteBranch codebase repo SyncMode.ShortCircuit Unmodified) \err ->
+                        Cli.returnEarly (Output.GitError err)
+                    ReadRemoteNamespaceShare repo -> importRemoteShareBranch repo
+              Cli.newBlock do
```
What is this `newBlock` here for?

ChrisPenner commented on unison-cli/src/Unison/CommandLine/Main.hs:
```diff
@@ -123,125 +117,117 @@ main dir welcome initialPath (config, cancelConfig) initialInputs runtime sbRunt
+          -- SIGINT
+          Left () -> do
+            hPutStrLn stderr "\nAborted."
+            loop
```
How did this even compile before the change? 🤔 
Is this one of those Non-Decreasing indentation quirks or something 😋

aryairani reviewed (changes requested):
> Reviewed 18 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/unisonweb/unison/3281)* status: 18 of 19 files reviewed, 14 unresolved discussions (waiting on @aryairani, @ChrisPenner, and @mitchellwrosen)

___
*[`parser-typechecker/package.yaml` line 175 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8yhICZAEIWBsm21IK6:-N8yhICZAEIWBsm21IK7:ba6olgu) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/parser-typechecker/package.yaml#L175)):*
> ```YAML
>   - TupleSections
>   - TypeApplications
>   - TypeFamilies
> ```

Where do we use the new extensions? Maybe I will find out as I keep reading.

Ah we said maybe overloaded label needed TypeFamilies.



_Code quote:_
```YAML
  - StandaloneDeriving
  - TupleSections
  - TypeApplications
  - TypeFamilies
```
___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 84 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Em-r1-84:-N8yoGN_3vX1KM1DklsZ:b-7ew2da) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Cli/Monad.hs#L84)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

This makes me a little nervous, I think it just means we'll have to be very careful about getting the root codebase in threads and such since it could change at any point without notice.

If an IORef allows us to use MonadUnliftIO though that might be nice?
</blockquote></details>

Per discussion, we'll switch to a pure state.
___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 99 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Em-r1-99:-N8yocWEEeeZEI8MuPFM:b1ch42b) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Cli/Monad.hs#L99)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

Is the lack of MonadUnliftIO going to be a problem?

It'd be great to have, but I assume it's somehow incompatible with the short-circuiting? Or is there some other reason?
</blockquote></details>

Per discussion, we have a safe resource-allocation strategy, so don't think lack of `MonadUnliftIO` is going to be a problem.
___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 102 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Em-r1-102:-N8yp2zk0vByYbrNXg1i:bwq3qsk) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Cli/Monad.hs#L102)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

I'm curious why `deriving via` rather than using  `ReaderT Env (ContT (ReturnType r) IO)` in the newtype directly?
</blockquote></details>

Travis: ergonomics something
___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 104 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Em-r1-104:-N8ypIUs94dDgmhtTLXv:baibp32) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Cli/Monad.hs#L104)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

Should probably clarify that `continue processing input` means short-circuiting the current user command and jumping straight to the next input, as it's written it's a bit fuzzy about whether it continues execution or skips things.
</blockquote></details>

Maybe update the comment to clarify that it's ending processing and continuing with the next input.
___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 166 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Em-r1-166:-N8ypko6A1CjbIH2tHgx:b-agcoay) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Cli/Monad.hs#L166)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

What was the rationale for using an overloaded label here as opposed to a regular ol' lens-as-a-value?

Is there some additional benefit? Otherwise I'd personally prefer to keep things simple and just use a regular lens 😄 , especially since we've already got a sprinkling of _both_ template haskell lenses and generic-lenses; adding another type might be confusing.

</blockquote></details>

Emulates a "currentPath" field that can be used like generic-lenses
___
*[`unison-cli/src/Unison/Codebase/Editor/Propagate.hs` line 243 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ysIvr9BHDJUNNP1_b:-N8ysIvr9BHDJUNNP1_c:begfwcx) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Codebase/Editor/Propagate.hs#L243)):*
> ```Haskell
>     pure noEdits
>   Just (initialTermEdits, initialTypeEdits) -> do
>     -- TODO: this can be removed once patches have term replacement
> ```

The "-- of type of `Referent -> Referent`" part of this comment was orphaned above when the comment was moved down. Also, not this PRs fault, but can we remember why?
___
*[`unison-cli/src/Unison/Codebase/Editor/HandleInput/AuthLogin.hs` line 179 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Eg-r1-179:-N8yt-pV-D5mzm6qCtiU:bomaoj6) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Codebase/Editor/HandleInput/AuthLogin.hs#L179)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

What's going on here? I assumed this was just a transplanted comment, but this string doesn't show up on trunk at all 🤔 
</blockquote></details>

To be deleted.
___
*[`unison-cli/src/Unison/CommandLine/Main.hs` line 223 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Ek-r1-223:-N8yrcsX9nW8pZLGALqx:b-knm3il) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/CommandLine/Main.hs#L223)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

How did this even compile before the change? 🤔 
Is this one of those Non-Decreasing indentation quirks or something 😋 
</blockquote></details>

This was a Github rendering issue.


<!-- Sent from Reviewable.io -->

ChrisPenner committed:
Merge branch 'trunk' into unison-cli-tech-debt-2

ChrisPenner committed:
Re-ormolu

mitchellwrosen committed:
put LoopState in a StateT layer rather than an IORef in Env

mitchellwrosen committed:
replace 'returnWith' with 'label'

mitchellwrosen committed:
Revert "replace 'returnWith' with 'label'"

This reverts commit 8455a19b26eb074109a47bbb233ce8aaf74cc77f.

mitchellwrosen committed:
rename with/withE to acquire/acquireE, and rework with/withE to take a callback

mitchellwrosen committed:
have 'time' take an inner action instead

mitchellwrosen committed:
doc tweaks

tstat committed:
add newBlock2

supports callCC-like tagged bailing but without extending the inner
continuation, and consequently without extending resource lifetimes

tstat committed:
runCli -> evalCli

tstat committed:
remove returnWith

tstat committed:
newBlock2 -> newBlock

tstat committed:
remove acquire*

tstat committed:
newBlock -> label

replace Either with isomorphic R to aid readability

tstat committed:
Breadcrumbs -> Label

tstat committed:
export only evalCli, renamed as runCli

tstat committed:
haddock tweaks

mitchellwrosen committed:
⅄ trunk → unison-cli-tech-debt-2

mitchellwrosen committed:
delete accidentally-committed files

aryairani reviewed (approved):
> Reviewed 3 of 6 files at r3, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/unisonweb/unison/3281)* status: 15 of 21 files reviewed, 15 unresolved discussions (waiting on @aryairani, @ChrisPenner, and @mitchellwrosen)

*[a discussion](https://reviewable.io/reviews/unisonweb/unison/3281#-N985eoQCihi50IPdCm-:-N985eoQCihi50IPdCm0:b-uwkowd) (no related file):*
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

___


<!-- Sent from Reviewable.io -->

mitchellwrosen committed:
add note about the boilerplate 'r' type variable

aryairani commented on unison-cli/src/Unison/Codebase/Editor/Propagate.hs:
```diff
@@ -234,17 +231,18 @@ debugMode = False
 propagate ::
-  Codebase IO Symbol Ann ->
-  Names -> -- TODO: this argument can be removed once patches have term replacement
   -- of type `Referent -> Referent`
```
```suggestion
```

aryairani commented on unison-cli/src/Unison/Codebase/Editor/Propagate.hs:
```diff
@@ -234,17 +231,18 @@ debugMode = False
+    Cli.respond PatchNeedsToBeConflictFree
     pure noEdits
   Just (initialTermEdits, initialTypeEdits) -> do
+    -- TODO: this can be removed once patches have term replacement
```
```suggestion
    -- TODO: this can be removed once patches have term replacement of type `Referent -> Referent`
```

aryairani committed:
Update unison-cli/src/Unison/Codebase/Editor/Propagate.hs

aryairani committed:
Update unison-cli/src/Unison/Codebase/Editor/Propagate.hs

aryairani reviewed (changes requested):
> Reviewed 2 of 3 files at r2, 1 of 6 files at r3, 2 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/unisonweb/unison/3281)* status: 20 of 21 files reviewed, 14 unresolved discussions (waiting on @aryairani, @ChrisPenner, and @mitchellwrosen)

___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 270 at r1](https://reviewable.io/reviews/unisonweb/unison/3281#-N8ygKtlk3WTKpYc66Em-r1-270:-N9CalOgCFEOnL4aa1gJ:bgt3p3q) ([raw file](https://github.com/unisonweb/unison/blob/e77fea71ed67039b98141e065b7268a8e40703d1/unison-cli/src/Unison/Cli/Monad.hs#L270)):*
<details><summary><i>Previously, ChrisPenner (Chris Penner) wrote…</i></summary><blockquote>

I think the fact that `newBlock` is the delimiter for all of `with`, `succeedWith` AND `time` might be a bit of a foot-gun;

E.g. what if I have a block like this:

```
do
  scratchFile <- with (withFile "scratch.u")
  doSomethingTimeIntensive
  useFile scratchFile
```

Then I want to time a block, I could see myself making this mistake (probably not in such a simple case, but in a real-world code situation it'd be easier to miss):

```
do
  newBlock do
    scratchFile <- with (withFile "scratch.u")
    doSomethingTimeIntensive
  useFile scratchFile
```

Now we free the file before using it :'( ;

You could make the same mistake using nested do-blocks too by returning the file outside of its scope, but it's considerably less likely IMO.
</blockquote></details>

We noted the compiler wouldn't allow this.
___
*[`unison-cli/src/Unison/Cli/Monad.hs` line 360 at r7](https://reviewable.io/reviews/unisonweb/unison/3281#-N9CaB28ELFQl9ItG0-8:-N9CaB28ELFQl9ItG0-9:bvrfubs) ([raw file](https://github.com/unisonweb/unison/blob/a90d4ad6cc0b5729d8c783ebfc9b9ce179c0b22c/unison-cli/src/Unison/Cli/Monad.hs#L360)):*
> ```Haskell
>   -- 'wrap'
>   default wrap :: s ~ t => s -> t
>   wrap = id
> ```

What is `wrap` and why don't we want haddocks to show for it? I guess because it's an internal implementation detail?

(Can be addressed later.)



_Code quote:_
```Haskell
class Label s t where
  wrap :: s -> t
  -- this default is only provided so that the haddocks don't show
  -- 'wrap'
  default wrap :: s ~ t => s -> t
  wrap = id
```


<!-- Sent from Reviewable.io -->

mitchellwrosen commented:
> What is wrap and why don't we want haddocks to show for it? I guess because it's an internal implementation detail?

The default implementation just prevents the line "Minimal complete definition: `wrap`" appearing in the haddocks. `wrap` itself isn't exported, because users don't write `Label` instances - the only two instances it will ever have are defined internally.

aryairani reviewed (approved):
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/unisonweb/unison/3281)* status: all files reviewed, 14 unresolved discussions (waiting on @ChrisPenner and @mitchellwrosen)



<!-- Sent from Reviewable.io -->

mitchellwrosen committed:
delete commented out block
